src/renderer/components/ContextMenu.tsx

```diff
@@ -228,32 +283,55 @@ export function ContextMenu(props: {
     }
   })
 
-  const items = props.items.filter(val => val !== false) as ContextMenuItem[]
-
   return (
-    <div
-      ref={menuRef}
-      className='dc-context-menu'
-      style={{
-        top: `${props.top}px`,
-        left: `${props.left}px`,
-      }}
-      role='menu'
-      tabIndex={-1}
-    >
-      {items.map((item, index) => (
+    <div>
+      {levelItems.map((level, levelIdx) => (
         <div
-          className='item'
-          onClick={() => {
-            didOpen.current = false
-            props.closeCallback()
-            item.action()
-          }}
+          ref={el => (menuLevelEls.current[levelIdx] = el as HTMLDivElement)}
+          key={levelIdx}
+          className='dc-context-menu'
+          role='menu'
           tabIndex={-1}
-          role='menuitem'
-          key={index}
+          style={{
+            top: `${props.top}px`,
+            left: `${props.left}px`,
+          }}
         >
-          {item.label}
+          {level.items.map((item, index) => (
+            <div
+              className={classNames({
+                item: true,
+                selected: index === openSublevels[levelIdx],
+              })}
+              onClick={(ev: React.MouseEvent) => {
+                if (item.subitems) {
+                  setSublevels(l => [...l.slice(0, levelIdx), index])
+                  keyboardFocus.current = 0
+                  ev.stopPropagation()
+                } else {
+                  setSublevels(l => l.slice(0, levelIdx))
+                  keyboardFocus.current = -1
+                  didOpen.current = false
+                  props.closeCallback()
+                  item.action()
+                }
+              }}
+              onMouseOver={() => {
+                if (item.subitems) {
+                  setSublevels(l => [...l.slice(0, levelIdx), index])
+                } else {
+                  setSublevels(l => l.slice(0, levelIdx))
+                }
+              }}
```

Comment:
How about moving the event handler functions outside of the React render method?

It's easier to read when the view- and control layer is separated, then I can read "what the component does" in one place - and "how it looks" in another. This also gives a nice opportunity to name our functions with something descriptive which helps me to understand what this is doing.

Reply:
Thanks for the feedback. I do agree with other things, but this one item I'd argue is better the way it is right now.

To provide a handler I'd have to do write a function producing another function just to pass additional parameters like this:
```
const makeMenuClickHandler = (
    item: ContextMenuItemActionable | ContextMenuItemExpandable,
    index: number,
    levelIdx: number
  ) => {
    return (ev: React.MouseEvent) => {
      if (item.subitems) {
        expandMenu(index, levelIdx)
        keyboardFocus.current = 0
...
```

And I feel it's a bit too much of an indirection. I do agree with adding helper functions expandSubMenu and collapseSubMenu, but that alone already add one step of indirection. What I like about current code it's in one place and it's clear about what's happening:
1. We iterate over a list of menu sublevels and then over the items to add these to DOM
2. Based on mouse input we either cut or extend this list. onMouseOver is pretty clean in that regard, onClick is more convoluted based on keyboard highlight behavior and how we handle the event, but I find it easier to digest in terms of how it all works and not what we were intending to do.

As a new contributor I'm obviously open to change it if it doesn't fit the general code style, I just feel that for now it's a better way to handle this exact case.